### ISSUE TITLE
Pytorch very slow to convert list of numpy arrays into tensors

### ISSUE THREAD

**Issue by lcswillems** [pytorch/pytorch#13918]
## 🐛 Bug

I compared the execution time of two codes.

Code 1:
```
import torch
import numpy as np

a = [np.random.randint(0, 10, size=(7, 7, 3)) for _ in range(100000)]
b = torch.tensor(np.array(a))
```
And code 2:
```
import torch
import numpy as np

a = [np.random.randint(0, 10, size=(7, 7, 3)) for _ in range(100000)]
b = torch.tensor(a)
```

The code 1 takes less than 1 second to execute (used `time`):
```
real    0m0,915s
user    0m0,808s
sys     0m0,330s
```

Whereas the code 2 takes 5 seconds:
```
real    0m6,057s
user    0m5,979s
sys     0m0,308s
```

## Expected behavior

I would expect code 2 to be as fast as code 1.

## Environment

 - PyTorch 0.4.1
 - Linux
 - OS (e.g., Linux):
 - Installed with `conda`
 - Python version: 3.6

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @mruberry @rgommers @heitorschueroff @VitalyFedyunin @ngimel

**Comment by SsnL** [comment]
You have an extra cast in the second case....

**Comment by lcswillems** [comment]
The case with the extra cast is faster!

**Comment by fmassa** [comment]
I believe in one of the cases we default to

**Comment by lcswillems** [comment]
Are you awaiting an answer of me?

**Comment by soumith** [comment]
@lcswillems no that tag was wrong, sorry about that.

**Comment by EthanGuan** [comment]
Yes, it is a bug, I think. I met the same problem, 200x faster in my code after I call a list of np.ndarray to np.ndarray.

**Comment by samedii** [comment]
This is a very unexpected performance issue and can take a long time to debug. It's quite likely that a lot of people don't find it at all

**Comment by JunkyByte** [comment]
Is this going to be fixed? I enjoy the 200x speedup I just got but I don't understand why I should spend 2 hours debugging a similar issue? List to tensor seems a reasonable conversion to me. While the workaround is convenient it is still a bug right? Thanks!

**Comment by paloha** [comment]
@ezyang how about a warning from the torch.tensor() applied to a list of np.arrays? You do not have to waste time fixing the issue and people will not have to waste time not knowing about it or debugging it.

**Comment by ezyang** [comment]
Yes, a warning seems very reasonable

**Comment by kurtamohler** [comment]
I will add a warning

**Comment by kylematoba** [comment]
Note as well that [the docs](https://pytorch.org/docs/stable/sparse.html) exhibit exactly this pattern.

**Comment by johngrabner** [comment]
The reported degradation is less than a factor of 10.  
My observations are much higher.

**masks = torch.as_tensor(np.array(masks), dtype=torch.uint8)**
time make torch.Size([0]) tensors 0:00:00.000086
time make torch.Size([0]) tensors 0:00:00.000080
time make torch.Size([1, 800, 800]) tensors 0:00:00.000231
time make torch.Size([5, 800, 800]) tensors 0:00:00.000780
time make torch.Size([9, 800, 800]) tensors 0:00:00.001698
time make torch.Size([10, 800, 800]) tensors 0:00:00.001889
time make torch.Size([12, 800, 800]) tensors 0:00:00.002627
time make torch.Size([16, 800, 800]) tensors 0:00:00.002116
time make torch.Size([17, 800, 800]) tensors 0:00:00.002246
time make torch.Size([18, 800, 800]) tensors 0:00:00.002294
time make torch.Size([18, 800, 800]) tensors 0:00:00.002379
time make torch.Size([19, 800, 800]) tensors 0:00:00.002634
time make torch.Size([19, 800, 800]) tensors 0:00:00.003685
time make torch.Size([19, 800, 800]) tensors 0:00:00.002485
time make torch.Size([19, 800, 800]) tensors 0:00:00.002507
time make torch.Size([19, 800, 800]) tensors 0:00:00.002554
time make torch.Size([19, 800, 800]) tensors 0:00:00.002546
time make torch.Size([19, 800, 800]) tensors 0:00:00.004128
time make torch.Size([20, 800, 800]) tensors 0:00:00.002839
time make torch.Size([20, 800, 800]) tensors 0:00:00.002621
time make torch.Size([20, 800, 800]) tensors 0:00:00.002867
time make torch.Size([21, 800, 800]) tensors 0:00:00.002730
time make torch.Size([34, 800, 800]) tensors 0:00:00.004624
time make torch.Size([35, 800, 800]) tensors 0:00:00.005996
time make torch.Size([39, 800, 800]) tensors 0:00:00.008200
time make torch.Size([52, 800, 800]) tensors 0:00:00.008860
time make torch.Size([76, 800, 800]) tensors 0:00:00.011913
time make torch.Size([90, 800, 800]) tensors 0:00:00.015364
time make torch.Size([98, 800, 800]) tensors 0:00:00.016864
time make torch.Size([109, 800, 800]) tensors 0:00:00.019374
time make torch.Size([115, 800, 800]) tensors 0:00:00.019694
time make torch.Size([120, 800, 800]) tensors 0:00:00.020211
time make torch.Size([131, 800, 800]) tensors 0:00:00.022619



**masks = torch.as_tensor(masks, dtype=torch.uint8)**
time make torch.Size([4, 800, 800]) tensors 0:00:02.640813
time make torch.Size([5, 800, 800]) tensors 0:00:03.367133
time make torch.Size([9, 800, 800]) tensors 0:00:06.209627
time make torch.Size([10, 800, 800]) tensors 0:00:06.893792
time make torch.Size([11, 800, 800]) tensors 0:00:07.557321
time make torch.Size([12, 800, 800]) tensors 0:00:07.895175
time make torch.Size([13, 800, 800]) tensors 0:00:09.071131
time make torch.Size([14, 800, 800]) tensors 0:00:09.232689
time make torch.Size([15, 800, 800]) tensors 0:00:10.146911
time make torch.Size([16, 800, 800]) tensors 0:00:10.655822
time make torch.Size([17, 800, 800]) tensors 0:00:11.257624
time make torch.Size([18, 800, 800]) tensors 0:00:11.938047
time make torch.Size([18, 800, 800]) tensors 0:00:11.920614
time make torch.Size([18, 800, 800]) tensors 0:00:12.392430
time make torch.Size([19, 800, 800]) tensors 0:00:12.494322
time make torch.Size([20, 800, 800]) tensors 0:00:13.983288
time make torch.Size([20, 800, 800]) tensors 0:00:13.230731
time make torch.Size([20, 800, 800]) tensors 0:00:13.075576
time make torch.Size([21, 800, 800]) tensors 0:00:14.487133
time make torch.Size([22, 800, 800]) tensors 0:00:15.219226
time make torch.Size([22, 800, 800]) tensors 0:00:15.239693
time make torch.Size([23, 800, 800]) tensors 0:00:15.021009
time make torch.Size([24, 800, 800]) tensors 0:00:16.449250
time make torch.Size([34, 800, 800]) tensors 0:00:22.142871
time make torch.Size([34, 800, 800]) tensors 0:00:22.357028
time make torch.Size([37, 800, 800]) tensors 0:00:25.969837
time make torch.Size([44, 800, 800]) tensors 0:00:28.506867
time make torch.Size([47, 800, 800]) tensors 0:00:30.702038
time make torch.Size([50, 800, 800]) tensors 0:00:33.150783
time make torch.Size([167, 800, 800]) tensors 0:01:53.621491

**Overall**
My degradations are a factor 3500x to 5300x times.

**Comment by GoingMyWay** [comment]
Same issue. Pytorch 1.9 has such an issue.
